In [1]:
%run imports
from src.functions import *
from src.classes import *
from src.plots import *
#195633320791780608, 195633325090480896


In [2]:
g166 = G166()

# Access and print the name and SkyCoord
print(g166.name)
print(g166.radius)
print(g166.SkyCoord)


G166
60 arcmin
<SkyCoord (ICRS): (ra, dec) in deg
    (81.625, 42.93333333)>


In [3]:
Star(195633320791780608).info
Star(195633325090480896).info

<Table length=1>
   RA_ICRS_1       DE_ICRS_1    ...                 SkyCoord                
      deg             deg       ...                deg,deg,pc               
    float64         float64     ...                 SkyCoord                
--------------- --------------- ... ----------------------------------------
 81.62564389768  42.93974983019 ... 81.62564389768,42.93974983019,3138.43115

In [3]:
coord = SkyCoord.from_name("SNR G166.0+04.3")
radius = 1*u.deg
# Gets all conical stars in the 1 deg region around G166 center:
# stars_in_region = get_stars_in_region(coord, radius)
stars_in_region = Table.read('data/stars_near_g166.ecsv', format='ascii.ecsv')
# stars_in_region
dist_mask_u = stars_in_region['rgeo']<5*u.kpc
dist_mask_l = stars_in_region['rgeo']>2.5*u.kpc
ruwe_mask = stars_in_region['RUWE']<1.4
plxq_mask = stars_in_region['Plx']/stars_in_region['e_Plx'] > 5
sir = stars_in_region[dist_mask_l & dist_mask_u & ruwe_mask & plxq_mask]
sir

<Table length=4749>
   RA_ICRS_1       DE_ICRS_1    ...                 SkyCoord                
      deg             deg       ...                deg,deg,pc               
    float64         float64     ...                 SkyCoord                
--------------- --------------- ... ----------------------------------------
 81.81732252743  42.29640171072 ... 81.81732252743,42.29640171072,3889.14697
 81.07020714699  42.30666259738 ... 81.07020714699,42.30666259738,3563.87891
 80.28360077127  43.07576252577 ... 80.28360077127,43.07576252577,3556.88159
 81.65634289596  43.01971193381 ... 81.65634289596,43.01971193381,3955.39429
 82.35367329679  43.11650908429 ... 82.35367329679,43.11650908429,4166.56787
 81.00295672003  42.41899671892 ... 81.00295672003,42.41899671892,2617.66113
 80.86337826456  42.39159306647 ... 80.86337826456,42.39159306647,3730.47852
 80.65165660438  42.40117278340 ...  80.65165660438,42.4011727834,2570.28589
 81.40775633236  42.83121275878 ... 81.40775633236,42.83121275878,3282.29395
            ...             ... ...                                      ...
 81.57998971292  43.70011438083 ...  81.57998971292,43.70011438083,2700.4104
 81.04189343161  42.67198971352 ... 81.04189343161,42.67198971352,2752.94629
 81.05560542368  42.61291651438 ... 81.05560542368,42.61291651438,2661.53247
 81.25241950321  43.41771544458 ... 81.25241950321,43.41771544458,2508.60254
 81.39458857461  43.52394377970 ...  81.39458857461,43.5239437797,2785.30127
 81.55926570834  43.69092229164 ... 81.55926570834,43.69092229164,2521.10547
 80.74086106362  42.63447602478 ... 80.74086106362,42.63447602478,2516.89185
 80.46493665119  43.38466993517 ... 80.46493665119,43.38466993517,2597.08716
 81.22518164710  42.54127927506 ...  81.2251816471,42.54127927506,4080.54614
 81.41594943215  42.29028243566 ... 81.41594943215,42.29028243566,3064.43994

# Optimise matching

- trace all stars back for 20 kyr
- create one config file for each star. 4749 config files needed. each config file performs the trace for one star with all possible matches.
- need to create these traceback config files.

In [11]:
star_lines = fs4giesler(sir)
star_lines #this is the input.

<Table length=4749>
TypeInput      RA           DE      ...    pmDE    e_pmDE        Source      
                                    ...  mas / yr mas / yr                   
  int64      str12        str12     ...  float64  float32        str21       
--------- ------------ ------------ ... --------- -------- ------------------
        1 05 27 16.157 42 17 47.046 ...    -0.170    0.019 195365594010029440
        1 05 24 16.850 42 18 23.985 ...    -0.218    0.020 195052027036610048
        1 05 21 08.064 43 04 32.745 ...    -0.550    0.016 207333812437636608
        1 05 26 37.522 43 01 10.963 ...     0.311    0.017 195645832035222400
        1 05 29 24.882 43 06 59.433 ...    -0.305    0.015 195474273863568000
        1 05 24 00.710 42 25 08.388 ...    -2.544    0.020 195241619775197824
        1 05 23 27.211 42 23 29.735 ...    -0.207    0.019 195243097243949184
        1 05 22 36.398 42 24 04.222 ...    -1.693    0.017 195232003349318144
        1 05 25 37.862 42 49 52.366 ...    -0.166    0.017 195642705299109760
      ...          ...          ... ...       ...      ...                ...
        1 05 26 19.198 43 42 00.412 ...    -3.189    0.070 207735821376033280
        1 05 24 10.054 42 40 19.163 ...    -1.785    0.071 195259761717193344
        1 05 24 13.345 42 36 46.499 ...    -0.112    0.074 195258593487895552
        1 05 25 00.581 43 25 03.776 ...    -3.495    0.073 207353461908793344
        1 05 25 34.701 43 31 26.198 ...    -1.088    0.076 207729464824481792
        1 05 26 14.224 43 41 27.320 ...    -3.850    0.078 207735821373397376
        1 05 22 57.807 42 38 04.114 ...    -3.863    0.068 195308483826167168
        1 05 21 51.585 43 23 04.812 ...    -1.745    0.077 207388130887037312
        1 05 24 54.044 42 32 28.605 ...    -0.730    0.146 195248667820329472
        1 05 25 39.828 42 17 25.017 ...    -0.556    0.059 195042990425340800

In [ ]:
execute_command('cd giessler_traceback ; examples/trace/traceback g166/trace.conf')

# testing

In [84]:
from astropy.table import Table
import astropy.units as u
import numpy as np

def find_neighbors_by_source_id(table, coord_column, source_column, source_id, max_sep, return_table=False):
    """
    Find stars within a specified angular separation from a given source ID in an astropy table.
    The output is sorted by separation, with the target source as the first row.

    Parameters:
    -----------
    table : astropy.table.Table
        The input table containing sky coordinates and source IDs.
    coord_column : str
        The column name of the SkyCoord objects in the table.
    source_column : str
        The column name of the source IDs in the table.
    source_id : int or str
        The source ID of the star to search around.
    max_sep : astropy.units.Quantity
        The maximum separation angle within which to find neighbors.
    return_table : bool, optional
        If True, returns a table of the neighbors with the target source as the first row.
        If False, returns a list of source IDs.

    Returns:
    --------
    neighbors : list of int or astropy.table.Table
        Either a list of source IDs of neighboring stars or a table of these neighbors,
        with the target star as the first row.
    """

    # Ensure max_sep is an astropy Quantity with angular units
    if not isinstance(max_sep, u.Quantity):
        raise ValueError("max_sep must be an astropy.units.Quantity with angular units (e.g., degrees).")

    # Locate the index of the target star with the given source_id
    target_indices = np.where(table[source_column] == source_id)[0]
    if len(target_indices) == 0:
        raise ValueError(f"Source ID {source_id} not found in the table.")
    target_index = target_indices[0]

    # Get the SkyCoord object of the star at the specified index
    target_coord = table[coord_column][target_index]

    # Calculate separations between the target and all other coordinates in the table
    separations = target_coord.separation(table[coord_column])

    # Find indices where separation is within the max_sep, including the target itself
    neighbor_indices = np.where(separations < max_sep)[0]

    # Sort the neighbors by separation
    sorted_indices = neighbor_indices[np.argsort(separations[neighbor_indices])]

    # Get the source IDs of the sorted neighboring stars
    neighbor_source_ids = table[source_column][sorted_indices].tolist()

    # Return results based on the specified return type
    if return_table:
        # Create a subtable with the sorted neighbors' source IDs
        neighbor_table = table[sorted_indices]
        return neighbor_table
    else:
        return neighbor_source_ids


In [86]:
neighbor_source_ids = find_neighbors_by_source_id(sir, "SkyCoord", "Source", source_id=195633325090480896, max_sep=3*u.arcmin, return_table=True)
# print("Neighbor source IDs:", neighbor_source_ids)
# fs4giesler(neighbor_source_ids)
neighbor_source_ids

<Table length=14>
   RA_ICRS_1       DE_ICRS_1    ...                 SkyCoord                
      deg             deg       ...                deg,deg,pc               
    float64         float64     ...                 SkyCoord                
--------------- --------------- ... ----------------------------------------
 81.62564389768  42.93974983019 ... 81.62564389768,42.93974983019,3138.43115
 81.62187230463  42.93624922588 ... 81.62187230463,42.93624922588,3241.66699
 81.61260055545  42.95209988890 ...  81.61260055545,42.9520998889,2658.02783
 81.64062967618  42.92659012262 ... 81.64062967618,42.92659012262,3680.09912
 81.60423028846  42.92918404038 ... 81.60423028846,42.92918404038,2703.84351
 81.59936703854  42.93817444048 ... 81.59936703854,42.93817444048,3332.52466
 81.59270938120  42.93048227897 ...  81.5927093812,42.93048227897,3031.43726
 81.58889547044  42.95540259001 ... 81.58889547044,42.95540259001,3142.18481
 81.60236243540  42.97840094338 ...  81.6023624354,42.97840094338,2601.40649
 81.57292207643  42.95940091546 ... 81.57292207643,42.95940091546,2735.39648
 81.59908095873  42.97934028246 ... 81.59908095873,42.97934028246,2612.82397
 81.60180901953  42.98306443331 ... 81.60180901953,42.98306443331,2893.06592
 81.61235710051  42.89340673209 ... 81.61235710051,42.89340673209,2846.47827
 81.68984257760  42.95144904211 ...  81.6898425776,42.95144904211,2514.20557